core/src/autogluon/core/scheduler/mo_asha_promotion.py

```diff
@@ -0,0 +1,213 @@
+        ret_key = None
+        if num_recorded >= int(round(1.0 / prom_quant)):
+            # Search for not yet promoted config in the top
+            # prom_quant fraction
+            def filter_pred(k, v):
+                return (not v[1]) and (config_key is None or k == config_key)
+
+            num_top = int(num_recorded * prom_quant)
+
+            evaluations = np.array([v[0] for _, v in recorded.items()])
+
+            if self.strategy == EPS_NET:
+                ranked_top = get_eps_net_ranking(evaluations, num_top)
+            elif self.strategy == NSGA_II:
+                ranked_top = get_nsga_ii_ranking(evaluations, num_top)
+            else:
+                raise ValueError("The provided selection strategy is invalid.")
+
+            items = list(recorded.items())
+            top_list = [items[i] for i in ranked_top]
+
+            try:
+                ret_key = next(k for k, v in top_list if filter_pred(k, v))
+            except StopIteration:
+                ret_key = None
+        return ret_key
+
+    @staticmethod
+    def _num_promotable_config(recorded, prom_quant):
+        raise NotImplementedError()
+
+
+# -----------------------------------------------------------------------------
+# Filtration techniques
+# -----------------------------------------------------------------------------
+
+def fast_nondominated_sort(values, num_samples):
+    # We assume a 2d np array of dim (n_candidates, n_objectives)
+    # This functions assumes a minimization problem. Implementation is based
+    # on the NSGA-II paper
+    assert len(values) > 0, "Need to provide at least 1 point."
+    assert len(values) >= num_samples, "Need at least enough points to meet \
+                                        num_samples."
+    domination_counts = np.zeros(len(values))
+    dominated_solutions = [[] for _ in range(len(values))]
+    fronts = [[]]
+    ranks = np.zeros(len(values))
+
+    for i, v1 in enumerate(values):
+        for j, v2 in enumerate(values):
+            if np.alltrue(v1 < v2):  # v1 dominates v2
+                dominated_solutions[i].append(j)
+            elif np.alltrue(v2 < v1):  # v2 dominates v1
+                domination_counts[i] += 1
+
+        if domination_counts[i] == 0:
+            ranks[i] = 1
+            fronts[0].append(i)
+
+    i = 0
+    n_selected = len(fronts[0])
+    while n_selected < num_samples:
+        assert len(fronts[i]) > 0, "Cannot select from empty front"
+        tmp = []
+        for j in fronts[i]:
+            for k in dominated_solutions[j]:
+                domination_counts[k] -= 1
+                if domination_counts[k] == 0:
+                    ranks[k] = i + 1
+                    tmp.append(k)
+        i += 1
+        n_selected += len(tmp)
+        fronts.append(tmp)
+    assert n_selected >= num_samples, "Could not assign enough samples"
+    return ranks, fronts
+
+
+def compute_eps_net(points: np.array, num_samples: int = None):
```

Comment:
Is points 1D here? It is (same name) 2D below. Not easy to understand. And please don't use len for numpy arrays, use size or shape. I am not even sure when len does!